File: EmojibleIOS/EmojibleIOS/DataSources/AssignmentDataSource.swift
//
//  AssignmentDataSource.swift
//  EmojibleIOS
//
//  Created by Furkan Yakal on 1.12.2020.


import Foundation
import Firebase

class AssignmentDataSource {
    
    let database = Firestore.firestore()
    var snapshotListener:ListenerRegistration?
    let notificationCenter = NotificationCenter.default
    var assignments: [AssignmentModel] = []
    var assignmentDataSourceIndices: [String] = []
    
    func startObservingAssignments(){
        guard let currentUser = AuthenticationManager.getInstance().currentUser else { return }
        let userDocRef = database.collection("Users").document(currentUser.uid)
        let assignmentsCollectionRef = userDocRef.collection("Assignments")
        snapshotListener = assignmentsCollectionRef.addSnapshotListener{ [unowned self] querySnapshot, error in
            guard let documents = querySnapshot?.documents else { return }
            assignments = documents.map{AssignmentModel(dictionary: $0.data())}
            assignmentDataSourceIndices = documents.map{$0.documentID}
            notificationCenter.post(name: .assignmentsChanged, object: nil, userInfo:["assignmentsChanged":assignments])
        }
    }
    
    func stopObservingAssignments(){
        guard let snapshotListener = snapshotListener else { return }
        snapshotListener.remove()
    }
    
    func editAssignment(oldAssignment: AssignmentModel, newValue: Any){
        guard let currentUser = AuthenticationManager.getInstance().currentUser else { return }
        var index: Int?
        for (i, document) in self.assignments.enumerated(){
            if document == oldAssignment{
                index = i
                break
            }
        }
        let uid = currentUser.uid
        database.collection("Users").document(uid).collection("Assignments").document(self.assignmentDataSourceIndices[index!]).setData(AssignmentModel(identifier: oldAssignment.identifier, value: newValue).dictionary)
        
    }
    
    func removeAssignment(assignment: AssignmentModel){
        guard let currentUser = AuthenticationManager.getInstance().currentUser else { return }
        var index: Int?
        for (i, document) in self.assignments.enumerated(){
            if document == assignment{
                index = i
                break
            }
        }
        let uid = currentUser.uid
        database.collection("Users").document(uid).collection("Assignments").document(self.assignmentDataSourceIndices[index!]).delete()
    }
    
    func writeAssignment(assignment: AssignmentModel){
        guard let currentUser = AuthenticationManager.getInstance().currentUser else { return }
        let uid = currentUser.uid
        self.database.collection("Users").document(uid).collection("Assignments").addDocument(data:assignment.dictionary)
    }
    
    private init(){}
    
    private static let instance = AssignmentDataSource()
    public static func getInstance() -> AssignmentDataSource{
        return .instance
    
    }
}

File: EmojibleAndroid/app/src/main/java/com/dji/emojibleandroid/dataSources/AssignmentDataSource.kt
package com.dji.emojibleandroid.dataSources

import com.dji.emojibleandroid.models.AssignmentModel
import com.dji.emojibleandroid.services.AuthenticationManager
import com.dji.emojibleandroid.services.Changes
import com.dji.emojibleandroid.services.NotificationCenter
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.ListenerRegistration
import com.google.firebase.firestore.ktx.firestore
import com.google.firebase.ktx.Firebase
import java.util.*
import kotlin.collections.HashMap

class AssignmentDataSource private constructor() {
    val database = Firebase.firestore
    var snapShotListener: ListenerRegistration? = null
    val notificationCenter = NotificationCenter.instance
    var assignments: List<AssignmentModel> = listOf()
    var assignmentDataSourceIndices: List<String> = listOf()

    fun startObservingAssignments() {
        val currentUser = FirebaseAuth.getInstance().currentUser.let { it } ?: return
        val userDocRef = database.collection("Users").document(currentUser.uid)
        val assignmentsCollectionRef = userDocRef.collection("Assignments")
        snapShotListener = assignmentsCollectionRef.addSnapshotListener { querySnapshot, error ->
            val documents = querySnapshot?.documents ?: return@addSnapshotListener
            assignments = documents.map { AssignmentModel(it.data as HashMap<String, Any>) }
            assignmentDataSourceIndices = documents.map { it.id }
            notificationCenter.post(
                Changes.assignmentsChanged,
                null,
                "assignmentsChanged" to assignments
            )
        }
    }

    fun stopObservingAssignments() {
        snapShotListener?.remove()
    }

    fun editAssignment(oldAssignment: AssignmentModel, newValue: Any) {
        val currentUser = AuthenticationManager.instance.currentUser
        var index: Int? = null
        assignments.forEachIndexed { i, document ->
            if (document == oldAssignment)
                index = i
            return
        }
        val uid = currentUser?.uid
        uid?.let {
            database.collection("Users").document(it).collection("Assignments")
                .document(assignmentDataSourceIndices[index!!]).set(
                    AssignmentModel(
                        identifier = oldAssignment.identifier,
                        v = newValue
                    ).dictionary
                )
        }
    }

    fun removeAssignment(assignment: AssignmentModel) {
        val currentUser = AuthenticationManager.instance.currentUser
        var index: Int? = null
        assignments.forEachIndexed { i, document ->
            if (document == assignment)
                index = i
            return
        }
        val uid = currentUser?.uid
        uid?.let {
            database.collection("Users").document(it).collection("Assignments")
                .document(assignmentDataSourceIndices[index!!]).delete()
        }
    }

    fun writeAssignment(assignment: AssignmentModel) {
        val currentUser = AuthenticationManager.instance.currentUser
        val uid = currentUser?.uid
        uid?.let {
            database.collection("Users").document(it).collection("Assignments")
                .add(assignment.dictionary)
        }
    }

    private object HOLDER {
        val INSTANCE = AssignmentDataSource()
    }

    companion object {
        val instance: AssignmentDataSource by lazy { HOLDER.INSTANCE }
    }
}
